docs: add guides (#430)
algolia/autocomplete: packages/website/docs/sending-algolia-insights-events.md

---
id: sending-algolia-insights-events
title: Sending Algolia Insights events
---

Learn how to send Algolia Insights events from your autocomplete menu.

:::note

Using this plugin requires an [Algolia](https://www.algolia.com/) application with the [Event Analytics](https://www.algolia.com/pricing/) feature enabled.

:::

> If you're using [Algolia indices](https://www.algolia.com/doc/faq/basics/what-is-an-index/) as [sources](sources) in your autocomplete, Algolia provides [Search Analytics](https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/) out-of-the-box. Search Analytics includes metrics like [top searches, top searches with no results, overall search counts, etc.](https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/#what-do-search-analytics-measure).

Reviewer: There are 2 dots at the end right now. Here's a suggestion:

```suggestion
If you're using [Algolia indices](https://www.algolia.com/doc/faq/basics/what-is-an-index/) as [sources](sources) in your autocomplete, Algolia provides [Search Analytics](https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/) out-of-the-box. Search Analytics includes metrics like [top searches, top searches with no results, overall search counts](https://www.algolia.com/doc/guides/getting-insights-and-analytics/search-analytics/out-of-the-box-analytics/#what-do-search-analytics-measure), etc.
```

        <img src={hit.image} alt={hit.name} width="40" height="40" />
      </div>
      <div className="aa-ItemContent">
        <div className="aa-ItemContentTitle">
          {highlightHit({ hit, attribute: "name" })}
        </div>
        <div className="aa-ItemContentDescription">
          {highlightHit({ hit, attribute: "description" })}
        </div>
      </div>
      <button
        className="aa-ItemActionButton aa-TouchOnly aa-ActiveOnly"
        type="button"
        title="Select"
      >
        <svg fill="currentColor" viewBox="0 0 24 24" width="20" height="20">
          <path d="M18.984 6.984h2.016v6h-15.188l3.609 3.609-1.406 1.406-6-6 6-6 1.406 1.406-3.609 3.609h13.172v-4.031z"></path>
        </svg>
      </button>
    </Fragment>
  );
}
```

This boilerplate assumes you want to insert the autocomplete into a DOM element with `autocomplete` as an [`id`](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id). You should change the [`container`](autocomplete-js/#container) to [match your markup](basic-options). Setting [`openOnFocus`](autocomplete-js/#openonfocus) to `true` ensures that the dropdown appears as soon as a user focuses the input.

> The autocomplete searches into an [Algolia index](https://www.algolia.com/doc/faq/basics/what-is-an-index/) of [BestBuy products](https://github.com/algolia/datasets/tree/master/ecommerce) using the [`getAlgoliaHits`](getAlgoliaHits) function. Refer to the example in the [Getting Started guide](getting-started) for more information.

Reviewer: To not overload users with information, I would keep it as "e-commerce":

```suggestion
The autocomplete searches into an [Algolia index](https://www.algolia.com/doc/faq/basics/what-is-an-index/) of [e-commerce products](https://github.com/algolia/datasets/tree/master/ecommerce) using the [`getAlgoliaHits`](getAlgoliaHits) function. Refer to the example in the [Getting Started guide](getting-started) for more information.
```

This snippet shows how to instantiate a plugin that sends a click event with "Product Selected from Autocomplete" as the `eventName` whenever a user selects an item.

```js title="index.js"
const appId = "latency";
const apiKey = "6be0576ff61c053d5f9a3225e2a90f76";
const searchClient = algoliasearch(appId, apiKey);
insightsClient("init", { appId, apiKey });

const algoliaInsightsPlugin = createAlgoliaInsightsPlugin({
  insightsClient,
  onSelect({ insights, insightsEvents }) {
    const events = insightsEvents.map((insightsEvent) => ({
      ...insightsEvent,
      eventName: 'Product Selected from Autocomplete',
    }));
    insights.clickedObjectIDsAfterSearch(...events);
  },
})
```

> If you're [using multiple different Algolia indices in the same autocomplete](adding-multiple-result-types), for example one for products and one for suggestions, you may want to use different `eventNames` for each section. You can do this conditionally based on the index name:

Reviewer: ```suggestion
If you're [using multiple different Algolia indices in the same autocomplete](adding-multiple-result-types), for example one for products and one for suggestions, you may want to use different `eventName`s for each section. You can do this conditionally based on the index name:
```

```js title="index.js"
const appId = "latency";
const apiKey = "6be0576ff61c053d5f9a3225e2a90f76";
const searchClient = algoliasearch(appId, apiKey);
insightsClient("init", { appId, apiKey });

const algoliaInsightsPlugin = createAlgoliaInsightsPlugin({
  insightsClient,
  onSelect({ insights, insightsEvents }) {
    const events = insightsEvents.map((insightsEvent) => {
      switch (insightsEvent.index) {
        case 'instant_search_demo_query_suggestions': {
          return {
            ...insightsEvent,
            eventName: 'Suggestion Selected from Autocomplete',
          };
        }
        case 'instant_search': {
          return {
            ...insightsEvent,
            eventName: 'Product Selected from Autocomplete',
          };
        }
        default: {
          return {
            ...insightsEvent,
            eventName: 'Item Selected from Autocomplete',
          };
        }
      }
    });
    insights.clickedObjectIDsAfterSearch(...events);
  },
});
```

> 

Reviewer: We should add one more section here, which could be called something like "Sending events from user interactions". If you want to send custom events from your templates (e.g., a conversion event from an "Add to cart" button), the Insights API is stored in the Context.

I created [an example in the playground](https://github.com/algolia/autocomplete.js/blob/3bc61d506a1fdde4091339ad91286cf74b9425b3/examples/js/app.tsx#L156-L161).

Author: Great idea! I've added a new section called "Sending events from templates" since the other sections could also be described by "Sending events from user actions".

Hope this title and section is what you were thinking of.

I've also moved this guide * after * the adding suggested searches, recent searches, and multiple results types guides since I think it makes more sense to learn this info after setting up your sources.

const searchClient = algoliasearch(appId, apiKey);
insightsClient("init", { appId, apiKey });

const algoliaInsightsPlugin = createAlgoliaInsightsPlugin({
  insightsClient,
  onSelect({ insights, insightsEvents }) {
    const events = insightsEvents.map((insightsEvent) => {
      switch (insightsEvent.index) {
        case 'instant_search_demo_query_suggestions': {
          return {
            ...insightsEvent,
            eventName: 'Suggestion Selected from Autocomplete',
          };
        }
        case 'instant_search': {
          return {
            ...insightsEvent,
            eventName: 'Product Selected from Autocomplete',
          };
        }
        default: {
          return {
            ...insightsEvent,
            eventName: 'Item Selected from Autocomplete',
          };
        }
      }
    });
    insights.clickedObjectIDsAfterSearch(...events);
  },
});
```

## Sending events from templates

Though the default Insights plugin doesn't send any conversion events, you may want to. For example, you may have created a [template](templates) with an "Add to cart" button.

![Autocomplete with add to cart button](/img/add-to-cart.jpg)

> If a user adds an item to their shopping cart directly from the autocomplete, you can send a conversion event directly from your template:

Reviewer: ```suggestion
If a user adds an item to their shopping cart directly from the autocomplete, you can send a conversion event from your template:
```

```jsx title="ProductItem.jsx"

> function ProductItem({ hit, insights }) {

Reviewer: We need to show users how to get the `insights` value, which is from the context in the templates:

```ts
autocomplete({
  // ...
  plugins: [algoliaInsightsPlugin],
  templates: {
    item({ item, state }) {
      return (
        <ProductItem
          hit={item}
          insights={state.context.algoliaInsightsPlugin.insights}
        />
      );
    },
  },
});
```

Author: Good point! Added here: https://github.com/algolia/autocomplete.js/pull/430/commits/348b8c7d316e08aecc836786e45be47c89af1139

> function ProductItem({ hit, insights }) {
>   return (
>     <Fragment>
>       <div className="aa-ItemIcon">
>         <img src={hit.image} alt={hit.name} width="40" height="40" />
>       </div>
>       <div className="aa-ItemContent">
>         <div className="aa-ItemContentTitle">
>           {snippetHit<ProductHit>({ hit, attribute: 'name' })}
>         </div>
>         <div className="aa-ItemContentDescription">
>           {snippetHit<ProductHit>({ hit, attribute: 'description' })}
>         </div>
>       </div>
>       <div className="aa-ItemActions">
>         <button
>           className="aa-ItemActionButton aa-TouchOnly aa-ActiveOnly"
>           type="button"
>           title="Select"
>         >
>           <svg fill="currentColor" viewBox="0 0 24 24" width="20" height="20">
>             <path d="M18.984 6.984h2.016v6h-15.188l3.609 3.609-1.406 1.406-6-6 6-6 1.406 1.406-3.609 3.609h13.172v-4.031z"></path>
>           </svg>
>         </button>
>         <button
>           className="aa-ItemActionButton"
>           type="button"
>           title="Add to cart"
>           onClick={(event) => {
>             event.preventDefault();
>             event.stopPropagation();
> 
>             insights.convertedObjectIDsAfterSearch({
>               eventName: 'Added to cart',
>               index: hit.__autocomplete_indexName,
>               objectIDs: [hit.objectID],
>               queryID: hit.__autocomplete_queryID,
>             });
>           }}
>         >
>           <svg
>             fill="none"
>             viewBox="0 0 24 24"
>             stroke="currentColor"
>             width="20"
>             height="20"
>           >
>             <path
>               strokeLinecap="round"
>               strokeLinejoin="round"
>               strokeWidth="2"
>               d="M3 3h2l.4 2M7 13h10l4-8H5.4M7 13L5.4 5M7 13l-2.293 2.293c-.63.63-.184 1.707.707 1.707H17m0 0a2 2 0 100 4 2 2 0 000-4zm-8 2a2 2 0 11-4 0 2 2 0 014 0z"
>             />
>           </svg>
>         </button>
>       </div>
>     </Fragment>
>   );
> }

Reviewer: We can remove unused code:

```suggestion
function ProductItem({ hit, insights }) {
  return (
    <Fragment>
      <div className="aa-ItemIcon">
        <img src={hit.image} alt={hit.name} width="40" height="40" />
      </div>
      <div className="aa-ItemContent">
        <div className="aa-ItemContentTitle">
          {snippetHit({ hit, attribute: 'name' })}
        </div>
        <div className="aa-ItemContentDescription">
          {snippetHit({ hit, attribute: 'description' })}
        </div>
      </div>
      <div className="aa-ItemActions">
        <button
          className="aa-ItemActionButton"
          type="button"
          title="Add to cart"
          onClick={(event) => {
            event.preventDefault();
            event.stopPropagation();
            insights.convertedObjectIDsAfterSearch({
              eventName: 'Added to cart',
              index: hit.__autocomplete_indexName,
              objectIDs: [hit.objectID],
              queryID: hit.__autocomplete_queryID,
            });
          }}
        >
          <svg
            viewBox="0 0 24 24"
            width="20"
            height="20"
            fill="none"
            stroke="currentColor"
          >
            <path
              d="M3 3h2l.4 2M7 13h10l4-8H5.4M7 13L5.4 5M7 13l-2.293 2.293c-.63.63-.184 1.707.707 1.707H17m0 0a2 2 0 100 4 2 2 0 000-4zm-8 2a2 2 0 11-4 0 2 2 0 014 0z"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
            />
          </svg>
        </button>
      </div>
    </Fragment>
  );
}
```

```

> This is possible since the Insights client is stored in [Context](context). Using it, you can build sending events into a [template](templates). For more information on the methods and event types you can send using the Insights client, consult the [Algolia Insights documenation](https://www.algolia.com/doc/api-client/methods/insights/).

Reviewer: ```suggestion
This is possible since the plugin stores the Insights API in [Context](context). Using it, you can send events from a [template](templates). For more information on the methods and event types you can send using the Insights API, consult the [Algolia Insights documentation](https://www.algolia.com/doc/api-client/methods/insights/).
```
